vault/rollback.go

```diff
@@ -158,7 +202,16 @@ func (m *RollbackManager) startOrLookupRollback(ctx context.Context, fullPath st
 	m.inflight[fullPath] = rs
 	rs.Add(1)
 	m.inflightAll.Add(1)
-	go m.attemptRollback(ctx, fullPath, rs, grabStatelock)
+
+	if workerChan != nil {
+		m.inflightLock.Unlock()
```

Comment:
Not sure I follow the subtleties here... I _think_ this is done because this add blocks until the channel is no longer full, and thus holding the lock could stall longer? But maybe we should just unlock in either case, as the `attemptRollback` in the direct goroutine case doesn't really need to hold the lock either (as otherwise, the m.attemptRollback would need to re-acquire the lock too.

Reply:
The issue is in the firstRun case, you aren't starting everything in a goroutine and then exiting the function, allowing the inflightLock to unlock.  Instead some of the jobs start and try get the inflightLock themselves and can't.  Then the workers fill up with jobs blocked in this way and we never get out of startRollback.  Deadlock.

Reply:
Right, but could we just unlock above on line 205, and not do the defer+return in the other branch?

Reply:
I think I was blindly worried about not still having the lock when we call attemptRollback in the non-worker case, but clearly we do nothing else afterward and that was meant to protect the map.